File: src/LC2023/LC238.java
package LC2023;

/**
 * @Author liutingfeng
 * @Date 2023/10/17 10:49
 */
public class LC238 {
    public int[] productExceptSelf(int[] nums) {
        // 1 2 3 4
        // 24 12 8 6
        // 1 1 2 6
        //
        int[] res = new int[nums.length];
        for (int i = 0; i < nums.length; i++) {
            if (i == 0) {
                res[i] = 1;
            } else {
                res[i] = res[i-1] * nums[i-1];
            }
        }
        int right = 1;
        for (int i = nums.length - 1; i >= 0; i--) {
            res[i] = res[i] * right;
            right *= nums[i];
        }
        return res;
    }

    public static void main(String[] args) {
        int[] nums = new int[]{1, 2, 3, 4};
        new LC238().productExceptSelf(nums);
    }
}

File: srcKotlin/LCKotlin/LC238.kt
package LCKotlin

class LC238 {

    fun productExceptSelf(nums: IntArray): IntArray {
        var res: IntArray = IntArray(nums.size, fun (_: Int): Int {
            return 1
        })
        //从左往右
        var t = 1
        for (i in 1 until nums.size) {
            t *= nums[i-1]
            res[i] = t
        }
        // 从右往左
        var rSum = 1
        for (i in nums.size - 2 downTo 0) {
            rSum *= nums[i+1]
            res[i] *= rSum
        }
        return res
    }
}